Add proposal: communication ingress flows matrix

sabinaaledort opened the pull request:
Communication ingress flows matrix of OpenShift Core and Operators: This enhancement allows automatically generate an accurate and up-to-date communication flows matrix that can be delivered to customers as part of product documentation for all ingress flows of OpenShift Core (multi-node) and Operators.

maiqueb committed:
SDN-4035: IPAM for VMs for OVN Kubernetes secondary networks

Provides three different alternatives:
- the "client" uses CNI args to express intent to the CNI IPAM plugin
- the feature's implementation happens in an external IPAM CNI plugin
- the "client" uses a declarative approch to express intent to the CNI plugin

The last option roughly consists of:
- IPAMClaim must have a status sub-resource
- KubeVirt creates the IPAMClaim object
- the "new" controller allocates an IP address and persists the CR
- KubeVirt monitors the IPAMClaim object for a VM until it as
  the IP address reported in the status
- KubeVirt templates the pod with that IP once available

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

cdoern committed:
MCO State Reporting

Introduce an enhancement discussing the MachineConfigNode introduction

Signed-off-by: Charlie Doern <cdoern@redhat.com>

maiqueb committed:
WIP: re-structure the doc

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
WIP: delete some more stuff

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
WIP: move previously chosen main alternative to yet another discarded one

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
WIP: remove redundant info

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
WIP: kubevirt only creates ipamClaims for nets wanting it

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
hot unplug considerations

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
KubeVirt must create the IPAMClaims to allow foreground deletion

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
indicate IPAMClaim name via network selection elements also for the discarded alternative

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
KubeVirt feat gate clarifications

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
remove hotunplug mermaid diagram

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

alanconway committed:
Added enhancements/cluster-logging/content-filter.md

andfasano committed:
simplify and unify day2 add nodes

andfasano committed:
Fixed reviewers/approvers list

andfasano committed:
linting fixes

andfasano committed:
vendoring as alternative

andfasano committed:
Refining cluster usage definition an oc integration

Tal-or committed:
enhancement: update APIs and changes done during implementation

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
enhancement: update implementation history

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
enhancement: update images

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

andfasano committed:
Minor syntax fixes from code review

Co-authored-by: Zane Bitter <zbitter@redhat.com>

andfasano committed:
additional minor changes

maiqueb committed:
link NPWG de-facto standard update proposal

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

andfasano committed:
added link to the integration notes

maiqueb committed:
make IPAMClaim examples consistent

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
clarify enhancement scope

While we do not care (business wise) about IPv6, explicitly calling that
out might be misleading.

Furthermore, we clarify this enhancement only targets secondary networks
of `localnet` or `layer2` topologies. The cluster default network /
routed secondary networks are **off** scope.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
reorder whereabouts is a bad idea arguments based on how important they are

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
clarify OVN-K role when allocating IP addresses

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

copejon committed:
microshift: expose audit log config values via user config file

Signed-off-by: Jon Cope <jcope@redhat.com>

RomanBednar committed:
add enhancement for Azure File cloning with azcopy

sabinaaledort committed:
Add proposal: communication ingress flows matrix

Communication ingress flows matrix of OpenShift and Operators:
This enhancement allows automatically generate an accurate and up-to-date
communication flows matrix that can be delivered to customers as part of
product documentation for all ingress flows of OpenShift (multi-node and
single node deployments) and Operators.

dhellmann committed:
this-week: 2024-02-02

openshift-merge-bot[bot] committed:
Merge pull request #1554 from dhellmann/twie-2024-02-02

this-week: 2024-02-02

dhellmann committed:
add step about filling in the metadata section at the top of the file

dhellmann committed:
be more explicit about looking for details for our different product topologies

dhellmann committed:
implementation history is tracked in jira now

dhellmann committed:
update the outline levels for some sections

dhellmann committed:
wording and whitespace changes

dhellmann committed:
add another example user story

openshift-merge-bot[bot] committed:
Merge pull request #1542 from RomanBednar/azure-storage-azcopy

STOR-1499: add enhancement for Azure File cloning with azcopy

maiqueb committed:
explicitly indicate allocation is lazy

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

slashpai committed:
MON-3514: Update windows node verification details for metrics-server

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

slashpai committed:
Update replace-prometheus-adapter-with-metrics-server.md

Co-authored-by: Simon Pasquier <spasquie@redhat.com>

dhellmann committed:
clarify template suggestion about implementation details

Co-authored-by: Aravindh Puthiyaparambil <aravindhp@users.noreply.github.com>

openshift-merge-bot[bot] committed:
Merge pull request #1557 from slashpai/metrics-server

MON-3514: Update windows node verification details for metrics-server

slashpai committed:
monitoring/OWNERS: update monitoring approvers

Signed-off-by: Jayapriya Pai <janantha@redhat.com>

openshift-merge-bot[bot] committed:
Merge pull request #1560 from slashpai/update_mon_owners

monitoring/OWNERS: update monitoring approvers

mrunalp committed:
Merge pull request #1532 from Tal-or/mixed-cpu-node-plugin

CNF-10784: Mixed cpu node plugin - Enhancement update

dhellmann committed:
this-week: 2024-02-09

openshift-merge-bot[bot] committed:
Merge pull request #1564 from dhellmann/twie-2024-02-09

this-week: 2024-02-09

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+
+This enhancement allows automatically generate an accurate and up-to-date 
+communication flows matrix that can be delivered to customers as part of 
+product documentation for all ingress flows of OpenShift Core (multi-node
```
not sure if "Openshift Core" is the right name (I think I invented it :-o ) 
maybe just "Openshift" is good enough?

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+## Motivation
+
+Security-conscious customers need OpenShift flows matrix for regulatory reasons 
+and/or to implement firewall rules to restrict traffic to the minimum set of 
```
do we want to mention can be on-node FW or external?

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+  generate an accurate, relevant, matrix for the running OpenShift cluster.
+
+- As a testing engineer, I want a machine-readable ingress traffic matrix 
+  so that I can craft and implement matching firewall rules.
```
this is to compare with tools such as nmap

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+
+- As an OpenShift cluster administrator, I want documentation on the expected 
+  flows of traffic incoming to every OpenShift installation so I can set up 
+  firewall rules to restrict traffic to the minimum required set of flows only.
```
this is to implement in FW such as nftables, NGFW, etc)

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+### Goals
+
+- Provide a mechanism to automatically generate an accurate and up-to-date 
+  OpenShift communication ingress flows matrix.
```
I think we also have a goal to make sure the documented OCP matrix does not regress (ie via origin test)

yuvalk replied:
this is probably our top priority ATM

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+- A communication matrix describing the expected flows of incoming traffic will 
+  be included in every OpenShift release documentation.
+
+- A new `oc` command will be added to generate an up-to-date communication
```
generate a current snapshot of known listening ports

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+  be included in every OpenShift release documentation.
+
+- A new `oc` command will be added to generate an up-to-date communication
+  matrix in a running cluster, `oc adm generate-commatrix`. #Sabina: what should be the command name?
```
`oc adm generate-network-listening-ports` ?
in any case I think it should include the word `network`
`oc adm network list-listening-ports` is even nicer, but I dont think we currently have anything else to put under `network`

and I agree it falls under `oc adm`

yuvalk replied:
oh, and even more important comment here, I dont think we're going to implement this part for now (but it is good to have it as part of the proposal)

sabinaaledort replied:
How about `oc adm communication-matrix generate` so we can later add more commands?

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+- A new option will be added to OpenShift web console to generate an up-to-date
+  communication matrix.
+
+### Workflow Description
```
I think we should also explain the flow of the test.
as well as the flow of looking up the doc (stupid but probably important to help understand the importance of keeping the doc up-to-date, a problem we have in other places)

yuvalk replied:
and which we are (hopefully) going to adress with:
a. doc automation
b. an origin test that will make sure the doc is up-to-date

sabinaaledort replied:
Should already be covered in workflow description

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+
+#### Variation and form factor considerations [optional] #Sabina: anything relevant here?
+
+How does this proposal intersect with Standalone OCP, Microshift and Hypershift?
```
I think we are leaving microshift and hypeshift out-of-scope here

yuvalk replied:
but I dont know enough about microshift

yuvalk replied:
for hypershift,
it's a nightmare for FW, even if it's on the node itself (and certainly if it's external)
the hosting cluster need to communicate with all clusters.
might be better if in the future it supports vlan tagging..
but for now, I think we better "ignore" it. say comm matrix is not support for hypershift

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,205 @@
+How does this proposal intersect with Standalone OCP, Microshift and Hypershift?
+
+If the cluster creator uses a standing desk, in step 1 above they can
+stand instead of sitting down.
```
this is a funny example, but probably better to remove it :-)

yuvalk reviewed (commented):
went all the way through implementation section.
I'm submitting this part and will continue the rest later

mainly looks good, these are all small comments :+1:

Miciah committed:
CONVENTIONS.md: s/well-know/well-known/

* CONVENTIONS.md (Metrics): Fix a typo ("well-know").

openshift-merge-bot[bot] committed:
Merge pull request #1555 from dhellmann/template-update-2024

update enhancement template

openshift-merge-bot[bot] committed:
Merge pull request #1517 from andfasano/day2-add-nodes

OCPSTRAT-784: Simplify and unify adding nodes on day 2 using agent-based installer approach

maiqueb committed:
de-facto standard update proposal was accepted

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

dhellmann committed:
this-week: 2024-02-16

openshift-merge-bot[bot] committed:
Merge pull request #1565 from dhellmann/twie-2024-02-16

this-week: 2024-02-16

openshift-merge-bot[bot] committed:
Merge pull request #1513 from alanconway/content-filter-enhancement

Added enhancements/cluster-logging/content-filter.md

maiqueb committed:
comply w/ new template

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

openshift-merge-bot[bot] committed:
Merge pull request #1456 from maiqueb/ipam-sdn-secondary-networks

SDN-4035: IPAM for VMs for OVN Kubernetes secondary networks

pmtk committed:
USHIFT-2186: Multus CNI for MicroShift (#1545)

* multus CNI for microshift

* review changes

* race condition risk between multus readiness and workload start

* add intro to race condition between MicroShift and other networking init procedures

* explicitly mention /opt/cni/bin

* NetPol and webhook are non-goals

* add goal about providing IPAMs and CNIs

* improve workflows

- added explicit RHEL (rpm) sections
- explicitly mention NAD before the App using NAD

* `vlan` -> `ipvlan`; images will be part of OCP payload

* reword origin of thick plugin

* add "longer startup of Pods" drawback

* fix trailing whitespace

* extra networks will not support K8s' Service API

* supportability: kubelet & crio logs

* add open question about --namespace-isolation

* Remove mention of `default-network` annotation

It's not intended to be used in normal mode of operation.

* mention dhcp server

* test day 2 installation

* alternative: Cluster Network Operator based operator

* fix multus.kubeconfig path

* OpenShift uses thick plugin since 4.14

* fix NAD example with nil gateway

* Q: logging to file?

* update template

* risk: NAD reusing MicroShift's CIDR

Miciah committed:
CONVENTIONS.md: Add section for TLS configuration

Document that components should use the TLS security profile from the
cluster APIServer config.

* CONVENTIONS.md (TLS configuration): New section.

openshift-merge-bot[bot] committed:
Merge pull request #1568 from Miciah/CONVENTIONS.md-add-section-for-TLS-configuration

CONVENTIONS.md: Add section for TLS configuration

dhellmann committed:
USHIFT-2348: microshift y-2 upgrades (#1562)

* USHIFT-2348: microshift y-2 upgrades

* add QE test plan

* feedback from architecture call

openshift-merge-bot[bot] committed:
Merge pull request #1490 from cdoern/machineconfigstate

MCO-811: MachineConfigNode introduction for MCO State Reporting

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+`EndpointSlices` group network endpoints together by unique combinations of 
+protocol, port number, and `Service` name and therefore can be used as the 
+source of truth for cluster traffic. For that to work, every application that 
+attracts ingress traffic running in the cluster should have a `Service` object,
```
they can also create endpointslices directly if they are doing "weird" stuff (like privileged parts of the platform that do not create a service, but are still listening on host net ports)
I'm basing this on wording found here
https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#management
that explain it's ok to create those

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+be filtered by `Service` type `NodePort` and `LoadBalancer` as there are also 
+`EndpointSlices` in the cluster for internal `Services`. The node hosted services 
+such as sshd, rpc, etc. are currently missing `EndpointSlices` and should be created 
+by the Machine API Operator during cluster deployment.
```
we'll probably be asked for *how* on this
and BTW: it's by Machine API, not Machine API Operator.
and I havent got to the detailed _how_ on this just yet, so I think we should add that to the open issues (that we are leaving this for a later enhancement - but we believe that the right place for it as part of adding/removing nodes)

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+such as sshd, rpc, etc. are currently missing `EndpointSlices` and should be created 
+by the Machine API Operator during cluster deployment.
+
+Sabina to Yuval: What k8s link you wanted to add here?
```
https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#management

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
````diff
@@ -0,0 +1,198 @@
+Each record describes a flow with the following information:
+```
+traffic-direction     Data flow direction (ingress/egress)
+protocol              Network protocol (TCP/UDP)
````
are we detecting udp/tcp only? what about other ip protocols (ICMP, SCTP, etc)?
what about non IP?

I think our MVP is TCP/UDP. then can extend to all IP
and that we should not cover non IP

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+traffic-direction     Data flow direction (ingress/egress)
+protocol              Network protocol (TCP/UDP)
+port-number           Flow port number
+service-name          OpenShift Service name
```
do we want to include more info such as container/pod name and maybe even the process name?

I'm thinking in the lines of how to make this easy for anyone who externally runs nmap to correlated with this.
so the protocol+port is the unique identifier, but it'll be weird (for them) if nmap shows "kube-rbac-proxy" and we say "MCO", so maybe we want to show that it's the kube-rbac-proxy that is inside MCO (just kube-rbac-proxy is also bad, because you cannot understand to what deployment it belongs)

yuvalk replied:
also, better be explicit here
is this is the service name? what if there is no service?
are we actually showing the deployment name?

let's also explain the logic involved in getting this info

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
````diff
@@ -0,0 +1,198 @@
+```
+
+For example,
+`ingress,TCP,18080,openshift-kni-infra-coredns-cnfdr8.telco5g.eng.rdu2.redhat.com-coredns,worker,true`
````
seems that this example does not correlate well with the above(??)
I seee direction, protocol,port,name,node-role,optional
(ie missing the namespace?)

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+`EndpointSlices` stand better for ingress traffic than for egress, in ingress traffic the 
+services are less dynamic and mostly stay up during the entire cluster lifetime.
+Supporting egress traffic might require changes in the API that should be reviewed and 
+agreed with the upstream Kubernetes community.
```
the egress risk - the risk is that when we'll get to do that part, which is currently out-of-scope, we might need to change the ingress part
especially considering the fact that we plan on bringing this upstream too

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+
+2. Should a new OpenShift object be introduced to act as the source of truth
+   for cluster traffic?
+
```
3. how to deal with privlieged (hostNet: true) workloads that can open new ports?
4. how to describe host services (probably by integrating with CAPI)
5. do we want to include both tools to get the declared flows and the actual currently running ones? maybe even offer a validation mode that will issue alerts on any miss. I'd also mention this whole thing might evolve into it's own operator

yuvalk commented on enhancements/network/communication-ingress-flows-matrix.md:
```diff
@@ -0,0 +1,198 @@
+- Unit tests coverage
+
+- E2E tests will be added to `openshift-tests`
+  - Validate the node open ports match the generated communication matrix
```
this is to validate against docs not just generated
and is the main tool to make sure we are not regressing

I think this is a crucial part of that proposal so much that I'd add more details to the implementation section

dhellmann committed:
update yaml to v3

dhellmann committed:
move owner file code to util package for reuse

dhellmann committed:
add prune-owners command

dhellmann committed:
remove approvers who are no longer org members

openshift-merge-bot[bot] committed:
Merge pull request #1576 from dhellmann/clean-up-staff-engineers

clean up owners file

arkadeepsen committed:
Update EP with the location of the DNSNameResolver controller code. Update the EP with the latest API details.

openshift-merge-bot[bot] committed:
Merge pull request #1551 from copejon/ushift-2196-audit-log-configuration

USHIFT-2196: User-facing audit-log configuration

dhellmann committed:
this-week: 2024-03-01

openshift-merge-bot[bot] committed:
Merge pull request #1580 from dhellmann/twie-2024-03-01

this-week: 2024-03-01

openshift-merge-bot[bot] committed:
Merge pull request #1579 from arkadeepsen/coredns-egressfirewall-integration

CFE-748: Update CoreDNS EgressFirewall integration enhancement proposal
